File: clients/elixir_client/lib/solidb/indexes_ops.ex
defmodule SoliDB.IndexesOps do
  @moduledoc """
  Advanced index operations API for SoliDB.
  """

  alias SoliDB.Client

  def rebuild(client, collection, index_name) do
    Client.send_command(client, "rebuild_index", %{database: Client.get_database(client), collection: collection, index_name: index_name})
  end

  def rebuild_all(client, collection) do
    Client.send_command(client, "rebuild_all_indexes", %{database: Client.get_database(client), collection: collection})
  end

  def hybrid_search(client, collection, query) do
    args = Map.merge(%{database: Client.get_database(client), collection: collection}, query)
    Client.send_command(client, "hybrid_search", args)
  end

  def analyze(client, collection, index_name) do
    Client.send_command(client, "analyze_index", %{database: Client.get_database(client), collection: collection, index_name: index_name})
  end

  def get_usage_stats(client, collection) do
    Client.send_command(client, "index_usage_stats", %{database: Client.get_database(client), collection: collection})
  end
end

File: clients/go-client/solidb/indexes_ops.go
package solidb

// IndexesClient provides access to advanced index management API
type IndexesClient struct {
	client *Client
}

// Rebuild rebuilds an index
func (i *IndexesClient) Rebuild(collection, indexName string) error {
	_, err := i.client.SendCommand("rebuild_index", map[string]interface{}{
		"database":   i.client.database,
		"collection": collection,
		"index_name": indexName,
	})
	return err
}

// RebuildAll rebuilds all indexes in a collection
func (i *IndexesClient) RebuildAll(collection string) error {
	_, err := i.client.SendCommand("rebuild_all_indexes", map[string]interface{}{
		"database":   i.client.database,
		"collection": collection,
	})
	return err
}

// HybridSearch performs hybrid search combining multiple search types
func (i *IndexesClient) HybridSearch(collection string, query map[string]interface{}) ([]interface{}, error) {
	args := map[string]interface{}{
		"database":   i.client.database,
		"collection": collection,
	}
	for k, v := range query {
		args[k] = v
	}
	res, err := i.client.SendCommand("hybrid_search", args)
	if err != nil {
		return nil, err
	}
	if slice, ok := res.([]interface{}); ok {
		return slice, nil
	}
	return nil, nil
}

// Analyze analyzes an index
func (i *IndexesClient) Analyze(collection, indexName string) (map[string]interface{}, error) {
	res, err := i.client.SendCommand("analyze_index", map[string]interface{}{
		"database":   i.client.database,
		"collection": collection,
		"index_name": indexName,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := res.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, nil
}

// GetUsageStats returns index usage statistics
func (i *IndexesClient) GetUsageStats(collection string) (map[string]interface{}, error) {
	res, err := i.client.SendCommand("index_usage_stats", map[string]interface{}{
		"database":   i.client.database,
		"collection": collection,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := res.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, nil
}
